File: B3642.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 1e6 + 5;
int n;
struct node {
    int l, r;
} a[N];

void front(int x)
{
    cout << x << ' ';
    if (a[x].l)
        front(a[x].l);
    if (a[x].r)
        front(a[x].r);
}

void mid(int x)
{
    if (a[x].l)
        mid(a[x].l);
    cout << x << ' ';
    if (a[x].r)
        mid(a[x].r);
}

void back(int x)
{
    if (a[x].l)
        back(a[x].l);
    if (a[x].r)
        back(a[x].r);
    cout << x << ' ';
}

int main()
{
    cin.tie(nullptr)->sync_with_stdio(false);

    cin >> n;
    for (int i = 1; i <= n; ++i) {
        cin >> a[i].l >> a[i].r;
    }

    front(1);
    cout << '\n';
    mid(1);
    cout << '\n';
    back(1);
    return 0;
}

File: B3642.py
class node:
    def __init__(self):
        self.l = None
        self.r = None


def front(x: int):
    print(x, end=" ")
    if t[x].l:
        front(t[x].l)
    if t[x].r:
        front(t[x].r)


def mid(x: int):
    if t[x].l:
        mid(t[x].l)
    print(x, end=" ")
    if t[x].r:
        mid(t[x].r)


def back(x: int):
    if t[x].l:
        back(t[x].l)
    if t[x].r:
        back(t[x].r)
    print(x, end=" ")


if __name__ == "__main__":
    n = int(input())
    t = [node() for _ in range(n + 1)]
    for i in range(1, n + 1):
        t[i].l, t[i].r = map(int, input().split())
    front(1)
    print()
    mid(1)
    print()
    back(1)
